entry/csp.js

```diff
@@ -0,0 +1,102 @@
+import { createHash } from 'crypto';
+import { parse } from 'node-html-parser';
+import { URL } from 'url';
+
+const scriptTypeIgnoreList = ['application/json', 'application/ld+json'];
+
+const defaultBaseName = 'http://relative-url';
+
+const hashScriptContents = (scriptContents) =>
+  createHash('sha256').update(scriptContents).digest('base64');
+
+export default function createCSPHandler({ extraHosts = [] } = {}) {
+  let tagReturned = false;
+  const hosts = new Set();
+  const shas = new Set();
+
+  const addScriptContents = (contents) => {
+    shas.add(hashScriptContents(contents));
+  };
+
+  const addScriptUrl = (src) => {
+    const { origin } = new URL(src, defaultBaseName);
+
+    if (origin !== defaultBaseName) {
+      hosts.add(origin);
+    }
+  };
+
+  extraHosts.forEach((host) => addScriptUrl(host));
+
+  const processScriptNode = (scriptNode) => {
+    const src = scriptNode.getAttribute('src');
+
+    if (src) {
+      addScriptUrl(src);
+    } else if (
+      !scriptTypeIgnoreList.includes(scriptNode.getAttribute('type'))
+    ) {
+      addScriptContents(scriptNode.firstChild.rawText);
+    }
+  };
+
+  const registerScript = (script) => {
+    if (tagReturned) {
+      throw new Error(
+        `Unable to register script. Content Security Policy already sent. Try registering scripts before calling flushHeadTags. Script: ${script.substr(
+          0,
+          30,
+        )}`,
+      );
+    }
+
+    const root = parse(script, { script: true });
```

Comment:
I'm just thinking now... this is on one of the hotest/critical paths of a server render. And most scripts will just be static strings re-used for every render.
It would be nice to have some way to pre-parse this HTML so we don't need to redo it everytime. I'm not sure memoization would work here as only some scripts may be static, some may change with every render.

I don't think there is anything to do now, but perhaps a future feature may allow you to pre-parse scripts like:
```ts
interface CsPValidAnalyticsScript {
  src?: string,
  content?: string,
}
const cspValidAnalyticsScript = generateCspValidScript(script);

registerScript(cspValidAnalyticsScript);
```

Reply:
Yeah I thought about this as well. Happy to add something later to improve the perf.